kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/resolver/ClusterResolutionResult.java

```diff
@@ -6,109 +6,197 @@
+     */
+    public boolean anyReferentResolvedRefsConditionsFalse() {
+        return findReferentsWithResolvedRefsFalse().findAny().isPresent();
     }
 
-    // using wildcards intentionally
-    @SuppressWarnings("java:S1452")
-    public Stream<LocalRef<?>> findReferentsWithStaleStatus() {
-        return referentsWithStaleStatus.stream();
+    private boolean isClusterStale() {
+        return Referent.from(cluster).isStale();
+    }
+
+    private boolean isClusterResolvedRefsFalse() {
+        return Referent.from(cluster).hasResolvedRefsFalseCondition();
     }
 
     // using wildcards intentionally
     @SuppressWarnings("java:S1452")
-    public Stream<LocalRef<?>> findResourcesWithResolvedRefsFalse(String kind) {
-        return findResourcesWithResolvedRefsFalse().filter(r -> kind.equals(r.getKind()));
+    private Stream<LocalRef<?>> findReferentsWithResolvedRefsFalse() {
+        return referentsMatching(Referent::hasResolvedRefsFalseCondition);
     }
 
-    ClusterResolutionResult addAllResourcesHavingResolvedRefsFalse(Set<LocalRef<?>> resolvedRefsFalse) {
-        return new ClusterResolutionResult(cluster, danglingReferences, Stream.concat(resourcesWithResolvedRefsFalse.stream(), resolvedRefsFalse.stream()).collect(
-                Collectors.toSet()), referentsWithStaleStatus);
+    private Stream<LocalRef<?>> referentsMatching(Predicate<Referent<?>> predicate) {
+        return allReferentResolutionResults().filter(i -> (i.referent() != null) && predicate.test(i.referent()))
+                .map(ResolutionResult::reference);
     }
 
-    ClusterResolutionResult addReferentsWithStaleStatus(Set<LocalRef<?>> referentsWithStaleStatus) {
-        return new ClusterResolutionResult(cluster, danglingReferences, resourcesWithResolvedRefsFalse,
-                Stream.concat(this.referentsWithStaleStatus.stream(), referentsWithStaleStatus.stream()).collect(
-                        Collectors.toSet()));
+    private Stream<ResolutionResult<?>> allReferentResolutionResults() {
+        Stream<ResolutionResult<?>> proxyResults = Stream.of(proxyResolutionResult);
+        Stream<ResolutionResult<?>> filterResults = filterResolutionResults.stream().map(Function.identity());
+        Stream<ResolutionResult<?>> serviceResults = Stream.of(serviceResolutionResult);
+        Stream<ResolutionResult<?>> ingressResults = ingressResolutionResults.stream()
+                .flatMap(i -> Stream.concat(Stream.of(i.ingressResolutionResult()), Optional.ofNullable(i.proxyResolutionResult()).stream()));
+        return Stream.of(proxyResults, filterResults, serviceResults, ingressResults).flatMap(Function.identity());
     }
 
     /**
-     * Describes the case where an entity A references an entity B but we can not find B.
+     * Describes the case where an entity A references an entity B, but we can not find B.
      * @param from
      * @param to
```

Comment:
Maybe rename `to` to `missing` or `absent`? I suspect that would make call sites more readable.

Reply:
done, `to` should be `referrer` as well.